File: api/app/schemas/dictamen.py
"""Pydantic models for `documents.ficha/partes/clausulas` and `analyses.dictamen` (v1.0).

Must stay in sync with `web/src/types/dictamen.ts` (see docs/HLD.md §3).
"""

from enum import Enum

from pydantic import BaseModel, Field

DICTAMEN_VERSION = "1.0"


class Nivel(str, Enum):
    critico = "critico"
    alto = "alto"
    medio = "medio"
    bajo = "bajo"
    informativo = "informativo"


class RecomendacionTipo(str, Enum):
    correccion = "correccion"
    tramite = "tramite"
    asesoria = "asesoria"


# --- documents.ficha / partes / clausulas -----------------------------------


class Ficha(BaseModel):
    plaza: str | None = None
    fecha: str | None = None
    cuantia: str | None = None
    forma_instrumental: str | None = None
    tipo_contrato: str | None = None
    rubro: str | None = None


class ParteDocumento(BaseModel):
    id: str
    nombre: str
    rol: str
    redacto: bool = False


class Clausula(BaseModel):
    id: str
    numero: str
    titulo: str | None = None
    texto: str


# --- analyses.dictamen (v1.0) ------------------------------------------------


class ArticuloCitado(BaseModel):
    articulo_id: str
    cuerpo: str
    numero: str
    inciso: str | None = None
    texto: str
    fuente_url: str | None = None


class ResumenPorNivel(BaseModel):
    critico: int = 0
    alto: int = 0
    medio: int = 0
    bajo: int = 0
    informativo: int = 0


class Resumen(BaseModel):
    hallazgos: int = 0
    omisiones: int = 0
    por_nivel: ResumenPorNivel = Field(default_factory=ResumenPorNivel)


class ParteDictamen(BaseModel):
    id: str
    nombre: str
    rol: str
    redacto: bool = False
    balance: int = 0
    a_favor: int = 0
    en_contra: int = 0
    lectura: str | None = None


class Hallazgo(BaseModel):
    id: str
    nivel: Nivel
    titulo: str
    clausula_id: str | None = None
    cita_textual: str | None = None
    fundamento: str
    articulos: list[ArticuloCitado] = Field(default_factory=list)
    redaccion_sustitutiva: str | None = None
    beneficia: str | None = None
    perjudica: str | None = None


class Omision(BaseModel):
    id: str
    nivel: Nivel
    titulo: str
    descripcion: str
    articulos: list[ArticuloCitado] = Field(default_factory=list)
    recomendacion: str | None = None


class Recomendacion(BaseModel):
    prioridad: int
    tipo: RecomendacionTipo
    accion: str


class Dictamen(BaseModel):
    version: str = DICTAMEN_VERSION
    indice_riesgo: int = Field(ge=0, le=100)
    nivel: Nivel
    confianza: float = Field(ge=0, le=1)
    resumen: Resumen
    sintesis: str
    partes: list[ParteDictamen] = Field(default_factory=list)
    hallazgos: list[Hallazgo] = Field(default_factory=list)
    omisiones: list[Omision] = Field(default_factory=list)
    recomendaciones: list[Recomendacion] = Field(default_factory=list)


def dictamen_json_schema() -> dict:
    return Dictamen.model_json_schema()

File: web/src/types/dictamen.ts
/**
 * Espejo TypeScript del esquema JSON descrito en docs/HLD.md §3.
 * Debe coincidir exactamente con api/app/schemas/dictamen.py.
 */

export type Rubro = "laboral" | "comercial" | "financiero" | "civil";

export type Nivel = "critico" | "alto" | "medio" | "bajo" | "informativo";

export type RecomendacionTipo = "correccion" | "tramite" | "asesoria";

// ---- documents.ficha / partes / clausulas ----------------------------------

export interface Ficha {
  plaza: string;
  fecha: string;
  cuantia: string;
  forma_instrumental: string;
  tipo_contrato: string;
  rubro: Rubro;
}

export interface Parte {
  id: string;
  nombre: string;
  rol: string;
  redacto: boolean;
}

export interface Clausula {
  id: string;
  numero: string;
  titulo: string;
  texto: string;
}

export interface DocumentoContrato {
  id: string;
  titulo: string;
  tipo_contrato: string;
  rubro: Rubro;
  ficha: Ficha;
  partes: Parte[];
  clausulas: Clausula[];
  texto?: string;
  ocr_status?: "pending" | "ready" | "failed";
  is_public?: boolean;
  created_at?: string;
}

// ---- analyses.dictamen (version 1.0) ---------------------------------------

export interface ResumenPorNivel {
  critico: number;
  alto: number;
  medio: number;
  bajo: number;
  informativo: number;
}

export interface Resumen {
  hallazgos: number;
  omisiones: number;
  por_nivel: ResumenPorNivel;
}

export interface ParteDictamen {
  id: string;
  nombre: string;
  rol: string;
  redacto: boolean;
  balance: number; // -100..100
  a_favor: number;
  en_contra: number;
  lectura: string;
}

export interface ArticuloCitado {
  articulo_id: string;
  cuerpo: string;
  numero: string;
  inciso?: string | null;
  texto: string;
  fuente_url: string;
}

export interface Hallazgo {
  id: string;
  nivel: Nivel;
  titulo: string;
  clausula_id: string;
  cita_textual: string;
  fundamento: string;
  articulos: ArticuloCitado[];
  redaccion_sustitutiva: string;
  beneficia?: string | null;
  perjudica?: string | null;
}

export interface Omision {
  id: string;
  nivel: Nivel;
  titulo: string;
  descripcion: string;
  articulos: ArticuloCitado[];
  recomendacion: string;
}

export interface Recomendacion {
  prioridad: number;
  tipo: RecomendacionTipo;
  accion: string;
}

export interface Dictamen {
  version: "1.0";
  indice_riesgo: number; // 0..100
  nivel: Nivel;
  confianza: number; // 0..1
  resumen: Resumen;
  sintesis: string;
  partes: ParteDictamen[];
  hallazgos: Hallazgo[];
  omisiones: Omision[];
  recomendaciones: Recomendacion[];
}

// ---- niveles: orden y presentación -----------------------------------------

export const NIVEL_ORDEN: Nivel[] = [
  "critico",
  "alto",
  "medio",
  "bajo",
  "informativo"
];

export const NIVEL_LABEL: Record<Nivel, string> = {
  critico: "Crítico",
  alto: "Alto",
  medio: "Medio",
  bajo: "Bajo",
  informativo: "Informativo"
};

export const NIVEL_COLOR: Record<Nivel, string> = {
  critico: "#B3261E",
  alto: "#D97706",
  medio: "#C9A227",
  bajo: "#2563A8",
  informativo: "#1E7A4C"
};
